Lottery reaching secondary limits

Seeing these logs in prod:

```

2025-01-14 12:19:00,002 INFO  [io.qua.git.lot.LotteryService] (executor-thread-1) Starting draw...
2025-01-14 12:19:02,460 INFO  [io.qua.git.lot.LotteryService] (executor-thread-1) Will draw for the following repositories: [quarkusio/quarkus(through GitHubInstallationRef[appSlug=quarkus-github-lottery, installationId=***]), quarkusio/quarkusio.github.io(through GitHubInstallationRef[appSlug=quarkus-github-lottery, installationId=***]), quarkusio/quarkus-github-lottery-reports(through GitHubInstallationRef[appSlug=quarkus-github-lottery, installationId=***])]
2025-01-14 12:19:02,460 INFO  [io.qua.git.lot.LotteryService] (executor-thread-1) Starting draw for repository quarkusio/quarkus(through GitHubInstallationRef[appSlug=quarkus-github-lottery, installationId=***])...
2025-01-14 12:19:05,224 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user yrodiere who wants to be notified on [MONDAY, WEDNESDAY, THURSDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:05,542 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user marko-bekhta who wants to be notified on [WEDNESDAY], because today is TUESDAY
2025-01-14 12:19:05,542 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user marko-bekhta because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:05,748 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user mbladel who wants to be notified on [WEDNESDAY], because today is TUESDAY
2025-01-14 12:19:05,748 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user mbladel because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:06,054 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping stewardship for user gsmet who wants to be notified on [MONDAY], because today is TUESDAY
2025-01-14 12:19:06,954 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping triage for user maxandersen who wants to be notified on [WEDNESDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:06,955 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user maxandersen who wants to be notified on [WEDNESDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:06,955 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user maxandersen because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:07,247 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping triage for user geoand who wants to be notified on [WEDNESDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:07,247 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user geoand who wants to be notified on [WEDNESDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:07,247 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping stewardship for user geoand who wants to be notified on [WEDNESDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:07,247 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user geoand because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:07,470 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping triage for user sberyozkin who wants to be notified on [MONDAY, WEDNESDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:07,471 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user sberyozkin who wants to be notified on [MONDAY, WEDNESDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:07,471 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user sberyozkin because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:08,239 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping triage for user cescoffier who wants to be notified on [MONDAY, WEDNESDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:08,239 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user cescoffier who wants to be notified on [MONDAY, WEDNESDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:08,239 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user cescoffier because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:08,531 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping triage for user jponge who wants to be notified on [WEDNESDAY], because today is TUESDAY
2025-01-14 12:19:08,531 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user jponge who wants to be notified on [WEDNESDAY], because today is TUESDAY
2025-01-14 12:19:08,531 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user jponge because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:08,824 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping triage for user aureamunoz who wants to be notified on [MONDAY, WEDNESDAY, THURSDAY], because today is TUESDAY
2025-01-14 12:19:08,824 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user aureamunoz who wants to be notified on [MONDAY, WEDNESDAY, THURSDAY], because today is TUESDAY
2025-01-14 12:19:08,824 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user aureamunoz because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:09,074 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping triage for user ozangunalp who wants to be notified on [MONDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:09,074 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user ozangunalp who wants to be notified on [MONDAY, FRIDAY], because today is TUESDAY
2025-01-14 12:19:09,074 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user ozangunalp because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:09,353 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping triage for user phillip-kruger who wants to be notified on [MONDAY], because today is TUESDAY
2025-01-14 12:19:09,354 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping maintenance for user phillip-kruger who wants to be notified on [FRIDAY], because today is TUESDAY
2025-01-14 12:19:09,354 DEBUG [io.qua.git.lot.dra.Participant] (executor-thread-1) Skipping user phillip-kruger because they participate in neither triage, maintenance nor stewardship
2025-01-14 12:19:12,163 ERROR [io.qua.git.lot.LotteryService] (executor-thread-1) Error drawing for repository quarkusio/quarkus(through GitHubInstallationRef[appSlug=quarkus-github-lottery, installationId=***]): org.kohsuke.github.GHException: Failed to retrieve https://api.github.com/search/issues?sort=updated&order=desc&q=repo%3Aquarkusio%2Fquarkus+is%3Aissue+is%3Aopen+label%3Atriage%2Fneeds-reproducer%2Ctriage%2Fneeds-feedback+label%3Aarea%2Fjakarta+updated%3A%3C2024-12-24T12%3A19%3A03.070504
	at org.kohsuke.github.GitHubPageIterator.fetch(GitHubPageIterator.java:157)
	at org.kohsuke.github.GitHubPageIterator.hasNext(GitHubPageIterator.java:93)
	at org.kohsuke.github.PagedSearchIterable$1.hasNext(PagedSearchIterable.java:111)
	at org.kohsuke.github.PagedIterator.fetch(PagedIterator.java:116)
	at org.kohsuke.github.PagedIterator.hasNext(PagedIterator.java:84)
	at java.base@21.0.5/java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1949)
	at java.base@21.0.5/java.util.stream.StreamSpliterators$WrappingSpliterator.lambda$initPartialTraversalState$0(StreamSpliterators.java:292)
	at java.base@21.0.5/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.fillBuffer(StreamSpliterators.java:206)
	at java.base@21.0.5/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.doAdvance(StreamSpliterators.java:161)
	at java.base@21.0.5/java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:298)
	at java.base@21.0.5/java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
	at io.quarkus.github.lottery.draw.Lottery$Draw.runSingleRound(Lottery.java:286)
	at io.quarkus.github.lottery.draw.Lottery.draw(Lottery.java:80)
	at io.quarkus.github.lottery.LotteryService.doDrawForRepository(LotteryService.java:109)
	at io.quarkus.github.lottery.LotteryService.drawForRepository(LotteryService.java:82)
	at io.quarkus.github.lottery.LotteryService.draw(LotteryService.java:66)
	at io.quarkus.github.lottery.LotteryService_ClientProxy.draw(Unknown Source)
	at io.quarkus.github.lottery.LotteryService_ScheduledInvoker_draw_6766887408be636b190d3a61d028cc72c29f541a.invokeBean(Unknown Source)
	at io.quarkus.scheduler.common.runtime.DefaultInvoker.invoke(DefaultInvoker.java:25)
	at io.quarkus.scheduler.common.runtime.DelegateInvoker.invokeDelegate(DelegateInvoker.java:29)
	at io.quarkus.scheduler.common.runtime.StatusEmitterInvoker.invoke(StatusEmitterInvoker.java:35)
	at io.quarkus.scheduler.common.runtime.DelegateInvoker.invokeDelegate(DelegateInvoker.java:29)
	at io.quarkus.scheduler.common.runtime.SkipConcurrentExecutionInvoker.invoke(SkipConcurrentExecutionInvoker.java:37)
	at io.quarkus.scheduler.common.runtime.DelegateInvoker.invokeDelegate(DelegateInvoker.java:29)
	at io.quarkus.scheduler.common.runtime.DelegateInvoker.invokeComplete(DelegateInvoker.java:36)
	at io.quarkus.scheduler.common.runtime.OffloadingInvoker$2.call(OffloadingInvoker.java:54)
	at io.quarkus.scheduler.common.runtime.OffloadingInvoker$2.call(OffloadingInvoker.java:51)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$4(ContextImpl.java:192)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextImpl$1.execute(ContextImpl.java:221)
	at io.vertx.core.impl.WorkerTask.run(WorkerTask.java:56)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:642)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@21.0.5/java.lang.Thread.runWith(Thread.java:1596)
	at java.base@21.0.5/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:896)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:872)
Caused by: org.kohsuke.github.HttpException: {
  "documentation_url": "https://docs.github.com/free-pro-team@latest/rest/overview/rate-limits-for-the-rest-api#about-secondary-rate-limits",
  "message": "You have exceeded a secondary rate limit. Please wait a few minutes before you try again. If you reach out to GitHub Support for help, please include the request ID ***."
}
	at org.kohsuke.github.GitHubConnectorResponseErrorHandler$1.onError(GitHubConnectorResponseErrorHandler.java:72)
	at org.kohsuke.github.GitHubClient.detectKnownErrors(GitHubClient.java:504)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:464)
	at org.kohsuke.github.GitHubPageIterator.fetch(GitHubPageIterator.java:146)
	... 42 more
```

--- a/.gitignore
+++ b/.gitignore
@@ -38,3 +38,4 @@ nb-configuration.xml
 
 # Local environment
 .env
+.quarkus

--- a/src/main/java/io/quarkus/github/lottery/config/DeploymentConfig.java
+++ b/src/main/java/io/quarkus/github/lottery/config/DeploymentConfig.java
@@ -1,0 +1,12 @@
+package io.quarkus.github.lottery.config;
+
+import io.smallrye.config.ConfigMapping;
+import io.smallrye.config.WithDefault;
+
+@ConfigMapping(prefix = "lottery")
+public interface DeploymentConfig {
+
+    @WithDefault("false")
+    boolean dryRun();
+
+}

--- a/src/main/java/io/quarkus/github/lottery/github/GitHubRepository.java
+++ b/src/main/java/io/quarkus/github/lottery/github/GitHubRepository.java
@@ -27,6 +27,7 @@ import java.util.function.Function;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
 
+import io.quarkus.github.lottery.config.DeploymentConfig;
 import org.kohsuke.github.GHDirection;
 import org.kohsuke.github.GHIssue;
 import org.kohsuke.github.GHIssueComment;
@@ -55,6 +56,7 @@ import io.smallrye.graphql.client.dynamic.api.DynamicGraphQLClient;
  */
 public class GitHubRepository implements AutoCloseable {
 
+    private final DeploymentConfig deploymentConfig;
     private final Clock clock;
     private final GitHubClientProvider clientProvider;
     private final GitHubConfigFileProvider configFileProvider;
@@ -66,8 +68,10 @@ public class GitHubRepository implements AutoCloseable {
     private GHRepository repository;
     private DynamicGraphQLClient graphQLClient;
 
-    public GitHubRepository(Clock clock, GitHubClientProvider clientProvider, GitHubConfigFileProvider configFileProvider,
+    public GitHubRepository(DeploymentConfig deploymentConfig, Clock clock,
+            GitHubClientProvider clientProvider, GitHubConfigFileProvider configFileProvider,
             MessageFormatter messageFormatter, GitHubRepositoryRef ref) {
+        this.deploymentConfig = deploymentConfig;
         this.clock = clock;
         this.clientProvider = clientProvider;
         this.configFileProvider = configFileProvider;
@@ -337,6 +341,12 @@ public class GitHubRepository implements AutoCloseable {
          */
         public void update(String topicSuffix, String markdownBody, boolean comment)
                 throws IOException {
+            if (deploymentConfig.dryRun()) {
+                Log.infof("[DRY RUN] Topic update:\n\tTopic:%s\n\tSuffix:%s\n\tBody:%s\n\tComment:%s",
+                        ref, topicSuffix, markdownBody, comment);
+                return;
+            }
+
             var dedicatedIssue = getDedicatedIssues().findFirst();
             if (ref.expectedSuffixStart() != null && !topicSuffix.startsWith(ref.expectedSuffixStart())
                     || ref.expectedSuffixStart() == null && !topicSuffix.isEmpty()) {

--- a/src/main/java/io/quarkus/github/lottery/github/GitHubService.java
+++ b/src/main/java/io/quarkus/github/lottery/github/GitHubService.java
@@ -4,6 +4,8 @@ import java.io.IOException;
 import java.time.Clock;
 import java.util.ArrayList;
 import java.util.List;
+
+import io.quarkus.github.lottery.config.DeploymentConfig;
 import jakarta.enterprise.context.ApplicationScoped;
 import jakarta.inject.Inject;
 
@@ -18,6 +20,8 @@ import org.kohsuke.github.GitHub;
 @ApplicationScoped
 public class GitHubService {
 
+    @Inject
+    DeploymentConfig deploymentConfig;
     @Inject
     Clock clock;
     @Inject
@@ -44,6 +48,6 @@ public class GitHubService {
     }
 
     public GitHubRepository repository(GitHubRepositoryRef ref) {
-        return new GitHubRepository(clock, clientProvider, configFileProvider, messageFormatter, ref);
+        return new GitHubRepository(deploymentConfig, clock, clientProvider, configFileProvider, messageFormatter, ref);
     }
 }

--- a/src/main/java/io/quarkus/github/lottery/util/RetryingIterator.java
+++ b/src/main/java/io/quarkus/github/lottery/util/RetryingIterator.java
@@ -1,0 +1,72 @@
+package io.quarkus.github.lottery.util;
+
+import java.io.InterruptedIOException;
+import java.io.UncheckedIOException;
+import java.util.Iterator;
+import java.util.function.Supplier;
+
+import org.kohsuke.github.GHException;
+import org.kohsuke.github.PagedIterator;
+
+import io.quarkus.logging.Log;
+
+// Workaround for https://github.com/hub4j/github-api/issues/2009
+class RetryingIterator<T> implements Iterator<T> {
+    private static final int MAX_RETRY = 3;
+
+    // Wait for unambiguously over one minute per GitHub guidance
+    private static final long DEFAULT_WAIT_MILLIS = 61 * 1000;
+
+    private final PagedIterator<T> delegate;
+
+    public RetryingIterator(PagedIterator<T> delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public boolean hasNext() {
+        return doWithRetry(delegate::hasNext);
+    }
+
+    @Override
+    public T next() {
+        return doWithRetry(delegate::next);
+    }
+
+    private <T> T doWithRetry(Supplier<T> action) {
+        RuntimeException rateLimitException = null;
+        for (int i = 0; i < MAX_RETRY; i++) {
+            if (rateLimitException != null) {
+                waitBeforeRetry(rateLimitException);
+            }
+            try {
+                return action.get();
+            } catch (RuntimeException e) {
+                if (isSecondaryRateLimitReached(e)) {
+                    if (rateLimitException == null) {
+                        rateLimitException = e;
+                    } else {
+                        rateLimitException.addSuppressed(e);
+                    }
+                } else {
+                    throw e;
+                }
+            }
+        }
+        throw rateLimitException;
+    }
+
+    private static boolean isSecondaryRateLimitReached(RuntimeException e) {
+        return e instanceof GHException
+                && e.getCause() != null && e.getCause().getMessage().contains("secondary rate limit");
+    }
+
+    private void waitBeforeRetry(RuntimeException e) {
+        Log.infof("GitHub API reached a secondary rate limit; waiting %s ms before retrying...", DEFAULT_WAIT_MILLIS);
+        try {
+            Thread.sleep(DEFAULT_WAIT_MILLIS);
+        } catch (InterruptedException ex) {
+            throw new UncheckedIOException((InterruptedIOException) new InterruptedIOException().initCause(e));
+        }
+    }
+}

--- a/src/main/java/io/quarkus/github/lottery/util/Streams.java
+++ b/src/main/java/io/quarkus/github/lottery/util/Streams.java
@@ -1,15 +1,19 @@
 package io.quarkus.github.lottery.util;
 
+import java.io.InterruptedIOException;
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 import java.util.NoSuchElementException;
+import java.util.Spliterator;
 import java.util.Spliterators;
 import java.util.function.BinaryOperator;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
+import org.kohsuke.github.GHException;
 import org.kohsuke.github.PagedIterable;
+import org.kohsuke.github.PagedIterator;
 
 public final class Streams {
 
@@ -63,10 +67,17 @@ public final class Streams {
     }
 
     public static <T> Stream<T> toStream(PagedIterable<T> iterable) {
-        return StreamSupport.stream(iterable.spliterator(), false);
+        return StreamSupport.stream(spliterator(iterable), false);
+    }
+
+    private static <T> Spliterator<T> spliterator(PagedIterable<T> iterable) {
+        var pagedIterator = iterable.iterator();
+        var workaroundIterator = new RetryingIterator<>(pagedIterator);
+        return Spliterators.spliteratorUnknownSize(workaroundIterator, 0);
     }
 
     public static <T> BinaryOperator<T> last() {
         return (first, second) -> second;
     }
+
 }

--- a/src/main/resources/application.properties
+++ b/src/main/resources/application.properties
@@ -11,9 +11,9 @@ quarkus.info.enabled=true
 %test.quarkus.log.category."io.quarkus.github.lottery".level=TRACE
 
 %dev.quarkus.scheduler.enabled=false
-%dev.quarkus.log.min-level=TRACE
-%dev.quarkus.log.category."io.quarkus.github.lottery".level=TRACE
-%dev.quarkus.log.category."org.kohsuke.github.GitHubClient".level=TRACE
+%dev.quarkus.log.min-level=FINEST
+%dev.quarkus.log.category."io.quarkus.github.lottery".level=FINEST
+%dev.quarkus.log.category."org.kohsuke.github.GitHubClient".level=FINEST
 
 %prod.quarkus.openshift.labels."app"=quarkus-github-lottery
 # Renew the SSL certificate automatically

